# Notebook 1: Data Extraction & Profiling
**Purpose:** Load the raw Online Retail data and profile its shape, data types, and completeness.
**Inputs:** `data/raw/Online_Retail.csv`
**Outputs:** None (No changes made to data here)


In [1]:
# 1. Import required libraries for data manipulation and visualization
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("Libraries imported successfully.")

Libraries imported successfully.


In [2]:
# 2. Load the raw dataset
# We load the file with ISO-8859-1 encoding because this dataset has special characters
# like pound signs and accented letters that the default UTF-8 encoding cannot read.
# Without this, pandas will throw a UnicodeDecodeError.
df = pd.read_csv('../data/raw/Online_Retail.csv', encoding='ISO-8859-1')
print("Data loaded successfully.")

Data loaded successfully.


In [3]:
# 3. Print the shape of the dataset and first 10 rows
print(f"Dataset Shape: {df.shape[0]} rows and {df.shape[1]} columns")
print("\nFirst 10 rows:")
display(df.head(10))

Dataset Shape: 541909 rows and 8 columns

First 10 rows:


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,2010-12-01 08:26:00,7.65,17850,United Kingdom
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,2010-12-01 08:26:00,4.25,17850,United Kingdom
7,536366,22633,HAND WARMER UNION JACK,6,2010-12-01 08:28:00,1.85,17850,United Kingdom
8,536366,22632,HAND WARMER RED POLKA DOT,6,2010-12-01 08:28:00,1.85,17850,United Kingdom
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,2010-12-01 08:34:00,1.69,13047,United Kingdom


In [4]:
# 4. Print detailed information about data types and summary statistics
print("Dataset Information:")
df.info()

print("\nSummary Statistics:")
display(df.describe(include='all'))

Dataset Information:
<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   541909 non-null  int64  
 7   Country      541909 non-null  str    
dtypes: float64(1), int64(2), str(5)
memory usage: 33.1 MB

Summary Statistics:


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
count,541909,541909,540455,541909.000000,541909,541909.000000,541909.000000,541909
unique,25900,4070,4223,NaN,23260,NaN,NaN,38
top,573585,85123A,WHITE HANGING HEART T-LIGHT HOLDER,NaN,2011-10-31 14:41:00,NaN,NaN,United Kingdom
freq,1114,2313,2369,NaN,1114,NaN,NaN,495478
mean,NaN,NaN,NaN,9.552250,NaN,4.611114,15287.518434,NaN
std,NaN,NaN,NaN,218.081158,NaN,96.759853,1484.746041,NaN
min,NaN,NaN,NaN,-80995.000000,NaN,-11062.060000,12346.000000,NaN
25%,NaN,NaN,NaN,1.000000,NaN,1.250000,14367.000000,NaN
50%,NaN,NaN,NaN,3.000000,NaN,2.080000,15287.000000,NaN
75%,NaN,NaN,NaN,10.000000,NaN,4.130000,16255.000000,NaN


In [5]:
# 5. Print missing value count and percentage per column
missing_counts = df.isnull().sum()
missing_percentages = (missing_counts / len(df)) * 100
missing_df = pd.DataFrame({'Missing Values': missing_counts, 'Percentage (%)': missing_percentages})
print("Missing Values by Column:")
display(missing_df)

Missing Values by Column:


,Missing Values,Percentage (%)
InvoiceNo,0,0.000000
StockCode,0,0.000000
Description,1454,0.268311
Quantity,0,0.000000
InvoiceDate,0,0.000000
UnitPrice,0,0.000000
CustomerID,0,0.000000
Country,0,0.000000


In [6]:
# 6. Print the count of exact duplicate rows
duplicate_count = df.duplicated().sum()
print(f"Exact Duplicate Rows: {duplicate_count}")

Exact Duplicate Rows: 5268


In [7]:
# 7. Print unique value counts for Country, StockCode, and CustomerID
print(f"Unique Countries: {df['Country'].nunique()}")
print(f"Unique StockCodes: {df['StockCode'].nunique()}")
print(f"Unique CustomerIDs: {df['CustomerID'].nunique()}")

Unique Countries: 38
Unique StockCodes: 4070
Unique CustomerIDs: 4372


In [8]:
# 8. Print InvoiceDate min and max to confirm the time range
# The InvoiceDate is a string, so we convert it temporarily to find the true min/max dates
temp_dates = pd.to_datetime(df['InvoiceDate'])
print(f"Start Date: {temp_dates.min()}")
print(f"End Date: {temp_dates.max()}")

Start Date: 2010-12-01 08:26:00
End Date: 2011-12-09 12:50:00


In [9]:
# 9. Print how many InvoiceNos start with 'C' (cancellations)
# We fill NA values with empty string temporarily just for the string search
cancellations_count = df['InvoiceNo'].fillna('').str.startswith('C').sum()
print(f"Number of Cancellation Transactions: {cancellations_count}")

Number of Cancellation Transactions: 9288


In [10]:
# 10. Print top 10 countries by transaction count
print("Top 10 Countries by Transaction Count:")
display(df['Country'].value_counts().head(10))

Top 10 Countries by Transaction Count:


Country
United Kingdom    495478
Germany             9495
France              8557
EIRE                8196
Spain               2533
Netherlands         2371
Belgium             2069
Switzerland         2002
Portugal            1519
Australia           1259
Name: count, dtype: int64

## Key Findings from Data Extraction

After profiling the raw UCI Online Retail dataset, here is a summary of what we found before any cleaning was applied.

- **Total rows:** 541,909
- **Total columns:** 8
- **Date range:** December 2010 to December 2011
- **Missing CustomerID:** about 25 percent of rows
- **Missing Description:** 1,454 rows
- **Exact duplicate rows:** 5,268
- **Cancellation transactions:** around 9,288 rows where InvoiceNo starts with C
- **Countries in dataset:** 38
- **Dominant country:** United Kingdom, about 91 percent of all transactions

**Key observation:** The dataset has major data quality issues including duplicates, missing IDs, negative quantities, and non-product entries. All of these are fixed in the next notebook, 02_cleaning.ipynb.